clusterloader2/api/validation.go

```diff
@@ -18,26 +18,187 @@ package api
 	}
+	return allErrs
+}
 
-	if errList.IsEmpty() {
-		return nil
+func (v *ConfigValidator) verifyStep(s *Step, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for i := range s.Phases {
+		allErrs = append(allErrs, v.verifyPhase(s.Phases[i], fldPath.Child("phases").Index(i))...)
+	}
+	for i := range s.Measurements {
+		allErrs = append(allErrs, v.verifyMeasurement(s.Measurements[i], fldPath.Child("measurements").Index(i))...)
+	}
+	isPhasesEmpty := len(s.Phases) == 0
+	isMeasurementsEmpty := len(s.Measurements) == 0
+	if isPhasesEmpty && isMeasurementsEmpty {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("measurements"), s.Measurements, "can't be empty with empty phases"))
+	}
+	if !isPhasesEmpty && !isMeasurementsEmpty {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("measurements"), s.Measurements, "can't be non-empty with non-empty phases"))
 	}
+	return allErrs
+}
 
-	return errList
+func (v *ConfigValidator) verifyPhase(p *Phase, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if p.NamespaceRange != nil {
+		allErrs = append(allErrs, v.verifyNamespaceRange(p.NamespaceRange, fldPath.Child("namespaceRange"))...)
+	}
+	if p.ReplicasPerNamespace < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("replicasPerNamespace"), p.ReplicasPerNamespace, "must be non-negative"))
+	}
+	for i := range p.ObjectBundle {
+		allErrs = append(allErrs, v.verifyObject(p.ObjectBundle[i], fldPath.Child("objectBundle").Index(i))...)
+	}
+	return allErrs
+}
+
+func (v *ConfigValidator) verifyObject(o *Object, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	errs := validation.IsDNS1123Subdomain(o.Basename)
+	for i := range errs {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("basename"), o.Basename, errs[i]))
+	}
+	if !v.fileExists(o.ObjectTemplatePath) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("objectTemplatePath"), o.ObjectTemplatePath, "file must exist"))
+	}
+	return allErrs
+}
+
+func (v *ConfigValidator) verifyNamespaceRange(nr *NamespaceRange, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if nr.Min < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("min"), nr.Min, "must be non-negative"))
+	}
+	if nr.Max < nr.Min {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("max"), nr.Max, "must be greater than min"))
+	}
+	return allErrs
+}
+
+func (v *ConfigValidator) verifyTuningSet(ts *TuningSet, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	tuningSetsNumber := 0
+	if ts.QPSLoad != nil {
+		tuningSetsNumber++
+		allErrs = append(allErrs, v.verifyQPSLoad(ts.QPSLoad, fldPath.Child("qpsLoad"))...)
+	}
+	if ts.RandomizedLoad != nil {
+		if tuningSetsNumber > 0 {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("randomizedLoad"), "may not specify more than 1 tuning set type"))
```

Comment:
This error should be on the `TuningSet` level (simply do that once at the end and unnested)